sqlair.go

```diff
@@ -0,0 +1,172 @@
+package sqlair
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/canonical/sqlair/internal/expr"
+)
+
+// Statement is a sql statement with valid sqlair DML.
+type Statement struct {
+	pe *expr.PreparedExpr
+}
+
+// Query stores a database query and iterates over the results.
+type Query struct {
+	pe  *expr.PreparedExpr
+	re  *expr.ResultExpr
+	err error
+	q   func() (*sql.Rows, error)
+}
+
+// Err returns the error, if any, that was encountered during iteration.
+// It should only be called after Close, or after Next has returned false (implicit close).
+func (q *Query) Err() error {
+	return q.err
+}
+
+type DB struct {
+	db *sql.DB
+}
+
+func NewDB(db *sql.DB) *DB {
+	return &DB{db: db}
+}
+
+// sqlDB returns the underlying database object.
+func (db *DB) sqlDB() *sql.DB {
+	return db.db
+}
+
+// Prepare checks that SQLair DML statements in a SQL query are well formed.
+// typeInstantiations must contain an instance of every type mentioned in the
+// SQLair DML of the query. These are used only for type information.
+func Prepare(query string, typeInstantiations ...any) (*Statement, error) {
+	parser := expr.NewParser()
+	parsedExpr, err := parser.Parse(query)
+	if err != nil {
+		return nil, err
+	}
+	preparedExpr, err := parsedExpr.Prepare(typeInstantiations...)
+	if err != nil {
+		return nil, err
+	}
+	return &Statement{pe: preparedExpr}, nil
+}
+
+func MustPrepare(query string, typeInstantiations ...any) *Statement {
+	s, err := Prepare(query, typeInstantiations...)
+	if err != nil {
+		panic(err)
+	}
+	return s
+}
+
+// Query takes a prepared SQLair Statement and returns a Query object for
+// iterating over the results.
+func (db *DB) Query(s *Statement, inputStructs ...any) (*Query, error) {
+	return db.QueryContext(s, context.Background(), inputStructs...)
+}
+
+func (db *DB) QueryContext(s *Statement, ctx context.Context, inputStructs ...any) (*Query, error) {
```

Comment:
It's probably worth noting why a user would want to pass a context to this function, perhaps something like
```
// QueryContext takes a prepared SQLAir statement, and returns a cancellable Query object for iterating over the results.
```
Also `QueryContext` feels like a getter, perhaps `QueryWithContext` would be clearer?

The above can also apply to `ExecContext` below.

Reply:
I've based these functions on these https://pkg.go.dev/database/sql#DB.Query . I've changed the comments to match the ones in the link a bit more.